File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-dividend-swaps-5-4/DividendPeriodPayment.cpp
// DividendPeriodPayment.cpp 
#include "DividendPeriodPayment.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

DividendPeriodPayment::DividendPeriodPayment(TiXmlNode* xmlNode)
: DividendPeriod(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //fixedStrikeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fixedStrikeNode = xmlNode->FirstChildElement("fixedStrike");

   if(fixedStrikeNode){fixedStrikeIsNull_ = false;}
   else{fixedStrikeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fixedStrikeNode , address : " << fixedStrikeNode << std::endl;
   #endif
   if(fixedStrikeNode)
   {
      if(fixedStrikeNode->Attribute("href") || fixedStrikeNode->Attribute("id"))
      {
          if(fixedStrikeNode->Attribute("id"))
          {
             fixedStrikeIDRef_ = fixedStrikeNode->Attribute("id");
             fixedStrike_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(fixedStrikeNode));
             fixedStrike_->setID(fixedStrikeIDRef_);
             IDManager::instance().setID(fixedStrikeIDRef_,fixedStrike_);
          }
          else if(fixedStrikeNode->Attribute("href")) { fixedStrikeIDRef_ = fixedStrikeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fixedStrike_ = boost::shared_ptr<PositiveDecimal>(new PositiveDecimal(fixedStrikeNode));}
   }

   //paymentDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* paymentDateNode = xmlNode->FirstChildElement("paymentDate");

   if(paymentDateNode){paymentDateIsNull_ = false;}
   else{paymentDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- paymentDateNode , address : " << paymentDateNode << std::endl;
   #endif
   if(paymentDateNode)
   {
      if(paymentDateNode->Attribute("href") || paymentDateNode->Attribute("id"))
      {
          if(paymentDateNode->Attribute("id"))
          {
             paymentDateIDRef_ = paymentDateNode->Attribute("id");
             paymentDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(paymentDateNode));
             paymentDate_->setID(paymentDateIDRef_);
             IDManager::instance().setID(paymentDateIDRef_,paymentDate_);
          }
          else if(paymentDateNode->Attribute("href")) { paymentDateIDRef_ = paymentDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { paymentDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(paymentDateNode));}
   }

   //valuationDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* valuationDateNode = xmlNode->FirstChildElement("valuationDate");

   if(valuationDateNode){valuationDateIsNull_ = false;}
   else{valuationDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- valuationDateNode , address : " << valuationDateNode << std::endl;
   #endif
   if(valuationDateNode)
   {
      if(valuationDateNode->Attribute("href") || valuationDateNode->Attribute("id"))
      {
          if(valuationDateNode->Attribute("id"))
          {
             valuationDateIDRef_ = valuationDateNode->Attribute("id");
             valuationDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(valuationDateNode));
             valuationDate_->setID(valuationDateIDRef_);
             IDManager::instance().setID(valuationDateIDRef_,valuationDate_);
          }
          else if(valuationDateNode->Attribute("href")) { valuationDateIDRef_ = valuationDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { valuationDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(valuationDateNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<PositiveDecimal> DividendPeriodPayment::getFixedStrike()
{
   if(!this->fixedStrikeIsNull_){
        if(fixedStrikeIDRef_ != ""){
             return boost::shared_static_cast<PositiveDecimal>(IDManager::instance().getID(fixedStrikeIDRef_));
        }else{
             return this->fixedStrike_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PositiveDecimal>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDate> DividendPeriodPayment::getPaymentDate()
{
   if(!this->paymentDateIsNull_){
        if(paymentDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(paymentDateIDRef_));
        }else{
             return this->paymentDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDate> DividendPeriodPayment::getValuationDate()
{
   if(!this->valuationDateIsNull_){
        if(valuationDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(valuationDateIDRef_));
        }else{
             return this->valuationDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-dividend-swaps-5-4/DividendPeriodPayment.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class DividendPeriodPayment : DividendPeriod
    {
        public DividendPeriodPayment(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode fixedStrikeNode = xmlNode.SelectSingleNode("fixedStrike");
            
            if (fixedStrikeNode != null)
            {
                if (fixedStrikeNode.Attributes["href"] != null || fixedStrikeNode.Attributes["id"] != null) 
                {
                    if (fixedStrikeNode.Attributes["id"] != null) 
                    {
                        fixedStrikeIDRef_ = fixedStrikeNode.Attributes["id"].Value;
                        PositiveDecimal ob = new PositiveDecimal(fixedStrikeNode);
                        IDManager.SetID(fixedStrikeIDRef_, ob);
                    }
                    else if (fixedStrikeNode.Attributes["href"] != null)
                    {
                        fixedStrikeIDRef_ = fixedStrikeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fixedStrike_ = new PositiveDecimal(fixedStrikeNode);
                    }
                }
                else
                {
                    fixedStrike_ = new PositiveDecimal(fixedStrikeNode);
                }
            }
            
        
            XmlNode paymentDateNode = xmlNode.SelectSingleNode("paymentDate");
            
            if (paymentDateNode != null)
            {
                if (paymentDateNode.Attributes["href"] != null || paymentDateNode.Attributes["id"] != null) 
                {
                    if (paymentDateNode.Attributes["id"] != null) 
                    {
                        paymentDateIDRef_ = paymentDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(paymentDateNode);
                        IDManager.SetID(paymentDateIDRef_, ob);
                    }
                    else if (paymentDateNode.Attributes["href"] != null)
                    {
                        paymentDateIDRef_ = paymentDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        paymentDate_ = new AdjustableOrRelativeDate(paymentDateNode);
                    }
                }
                else
                {
                    paymentDate_ = new AdjustableOrRelativeDate(paymentDateNode);
                }
            }
            
        
            XmlNode valuationDateNode = xmlNode.SelectSingleNode("valuationDate");
            
            if (valuationDateNode != null)
            {
                if (valuationDateNode.Attributes["href"] != null || valuationDateNode.Attributes["id"] != null) 
                {
                    if (valuationDateNode.Attributes["id"] != null) 
                    {
                        valuationDateIDRef_ = valuationDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(valuationDateNode);
                        IDManager.SetID(valuationDateIDRef_, ob);
                    }
                    else if (valuationDateNode.Attributes["href"] != null)
                    {
                        valuationDateIDRef_ = valuationDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        valuationDate_ = new AdjustableOrRelativeDate(valuationDateNode);
                    }
                }
                else
                {
                    valuationDate_ = new AdjustableOrRelativeDate(valuationDateNode);
                }
            }
            
        
        }
        
    
        #region FixedStrike_
        private PositiveDecimal fixedStrike_;
        public PositiveDecimal FixedStrike_
        {
            get
            {
                if (this.fixedStrike_ != null)
                {
                    return this.fixedStrike_; 
                }
                else if (this.fixedStrikeIDRef_ != null)
                {
                    fixedStrike_ = IDManager.getID(fixedStrikeIDRef_) as PositiveDecimal;
                    return this.fixedStrike_; 
                }
                else
                {
                      return this.fixedStrike_; 
                }
            }
            set
            {
                if (this.fixedStrike_ != value)
                {
                    this.fixedStrike_ = value;
                }
            }
        }
        #endregion
        
        public string fixedStrikeIDRef_ { get; set; }
        #region PaymentDate_
        private AdjustableOrRelativeDate paymentDate_;
        public AdjustableOrRelativeDate PaymentDate_
        {
            get
            {
                if (this.paymentDate_ != null)
                {
                    return this.paymentDate_; 
                }
                else if (this.paymentDateIDRef_ != null)
                {
                    paymentDate_ = IDManager.getID(paymentDateIDRef_) as AdjustableOrRelativeDate;
                    return this.paymentDate_; 
                }
                else
                {
                      return this.paymentDate_; 
                }
            }
            set
            {
                if (this.paymentDate_ != value)
                {
                    this.paymentDate_ = value;
                }
            }
        }
        #endregion
        
        public string paymentDateIDRef_ { get; set; }
        #region ValuationDate_
        private AdjustableOrRelativeDate valuationDate_;
        public AdjustableOrRelativeDate ValuationDate_
        {
            get
            {
                if (this.valuationDate_ != null)
                {
                    return this.valuationDate_; 
                }
                else if (this.valuationDateIDRef_ != null)
                {
                    valuationDate_ = IDManager.getID(valuationDateIDRef_) as AdjustableOrRelativeDate;
                    return this.valuationDate_; 
                }
                else
                {
                      return this.valuationDate_; 
                }
            }
            set
            {
                if (this.valuationDate_ != value)
                {
                    this.valuationDate_ = value;
                }
            }
        }
        #endregion
        
        public string valuationDateIDRef_ { get; set; }
        
    
        
    
    }
    
}

